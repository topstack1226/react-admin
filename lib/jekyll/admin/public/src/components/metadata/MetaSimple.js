import React, { Component, PropTypes } from 'react';
import ContentEditable from 'react-contenteditable';
import _ from 'underscore';
import DateTimePicker from 'react-widgets/lib/DateTimePicker';
import dateformat from 'dateformat';
import moment from 'moment';
import momentLocalizer from 'react-widgets/lib/localizers/moment';
import 'react-widgets/dist/css/react-widgets.css';

momentLocalizer(moment);

export class MetaSimple extends Component {

  handleEditableChange(e) {
    const { nameAttr, fieldValue, updateFieldValue } = this.props;
    let currentValue = e.target.value;
    if (fieldValue != currentValue && currentValue != '') {
      updateFieldValue(nameAttr, currentValue);
    }
  }

  handleDatepickerChange(value, dateStr) {
    const { nameAttr, fieldValue, updateFieldValue } = this.props;
    if (fieldValue != dateStr && dateStr != '') {
      let formatted = dateformat(dateStr, "yyyy-mm-dd HH:MM:ss");
      updateFieldValue(nameAttr, formatted.toString());
    }
  }

  renderEditable() {
    const { fieldValue } = this.props;
    return (
      <ContentEditable
        onChange={(e) => this.handleEditableChange(e)}
        className="field value-field"
        html={fieldValue} />
    );
  }

  renderDatepicker() {
    const { fieldValue } = this.props;
    let dateValue = (new Date(fieldValue) == 'Invalid Date') ?
      null :
      new Date(fieldValue);
    return (
      <DateTimePicker
        onChange={(v, d) => this.handleDatepickerChange(null, d)}
        className="date-field"
        editFormat={"MMM DD, YYYY HH:MM"}
        format={"MMM DD YYYY HH:MM"}
        defaultValue={dateValue} />
    );
  }

  render() {
    const { fieldKey } = this.props;
    return (
      <div className="meta-value">
        {
          fieldKey == 'date' && this.renderDatepicker()
        }
        {
          fieldKey != 'date' && this.renderEditable()
        }
      </div>
    );
  }

}

MetaSimple.propTypes = {
  parentType: PropTypes.string.isRequired,
  fieldKey: PropTypes.string.isRequired,
  fieldValue: PropTypes.any.isRequired,
  updateFieldValue: PropTypes.func.isRequired,
  nameAttr: PropTypes.any.isRequired
};

export default MetaSimple;
