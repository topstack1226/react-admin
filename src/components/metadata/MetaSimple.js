import React, { Component, PropTypes } from 'react';
import TextareaAutosize from 'react-textarea-autosize';
import _ from 'underscore';
import DateTimePicker from 'react-widgets/lib/DateTimePicker';
import moment from 'moment';
import momentLocalizer from 'react-widgets/lib/localizers/moment';
import 'react-widgets/dist/css/react-widgets.css';

momentLocalizer(moment);

export class MetaSimple extends Component {

  handleEditableChange(e) {
    const { nameAttr, fieldValue, updateFieldValue } = this.props;
    updateFieldValue(nameAttr, e.target.value);
  }

  handleDatepickerChange(date, dateStr) {
    const { nameAttr, fieldValue, updateFieldValue } = this.props;
    let formatted = moment(date).format("YYYY-MM-DD hh:mm:ss");
    updateFieldValue(nameAttr, formatted);
  }

  renderEditable() {
    const { fieldValue } = this.props;
    return (
      <TextareaAutosize
        onChange={(e) => this.handleEditableChange(e)}
        className="field value-field"
        defaultValue={fieldValue} />
    );
  }

  renderDatepicker() {
    const { fieldValue } = this.props;
    let dateValue = (new Date(fieldValue) == 'Invalid Date') ? null : new Date(fieldValue);
    return (
      <DateTimePicker
        onChange={(v, d) => this.handleDatepickerChange(v, d)}
        className="date-field"
        defaultValue={dateValue} />
    );
  }

  render() {
    const { fieldKey } = this.props;
    return (
      <div className="meta-value">
        {fieldKey == 'date' && this.renderDatepicker()}
        {fieldKey != 'date' && this.renderEditable()}
      </div>
    );
  }

}

MetaSimple.propTypes = {
  parentType: PropTypes.string.isRequired,
  fieldKey: PropTypes.string.isRequired,
  fieldValue: PropTypes.any.isRequired,
  updateFieldValue: PropTypes.func.isRequired,
  nameAttr: PropTypes.any.isRequired
};

export default MetaSimple;
