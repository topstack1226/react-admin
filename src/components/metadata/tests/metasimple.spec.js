import React from 'react';
import { Link } from 'react-router';
import expect from 'expect';
import { mount } from 'enzyme';

import MetaSimple from '../MetaSimple';

const defaultProps = {
  parentType: 'top',
  fieldKey: 'layout',
  fieldValue: 'page',
  nameAttr: 'metadata["layout"]'
};

function setup(props = defaultProps) {
  const actions = {
    updateFieldValue: expect.createSpy()
  };

  let component = mount(
    <MetaSimple {...props} {...actions} />
  );

  return {
    component,
    editable: component.find('.value-field'),
    datepicker: component.find('.date-field'),
    actions,
    props
  };
}

describe('Components::MetaSimple', () => {
  it('should render editable if field key is not called date', () => {
    const { editable, datepicker } = setup();
    expect(editable.node).toExist();
    expect(datepicker.node).toNotExist();
  });

  it('should render datepicker if field key is called date', () => {
    const { editable, datepicker } = setup(Object.assign({}, defaultProps, {
      fieldKey: 'date'
    }));
    expect(datepicker.node).toExist();
    expect(editable.node).toNotExist();
  });

  it('should call updateFieldValue when the input is changed', () => {
    const { actions, editable } = setup();
    editable.simulate('change');
    expect(actions.updateFieldValue).toHaveBeenCalled();
  });
});
