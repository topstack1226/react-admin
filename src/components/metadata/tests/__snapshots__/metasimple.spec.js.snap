// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::MetaSimple renders datepicker if field key is called date 1`] = `
<div
  className="meta-value"
>
  <div
    className="date-field rw-datetimepicker rw-has-both rw-widget"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    tabIndex="-1"
  >
    <input
      aria-busy={false}
      aria-describedby={undefined}
      aria-disabled={undefined}
      aria-expanded={false}
      aria-haspopup={true}
      aria-labelledby={undefined}
      aria-owns="rw_1_cal rw_1_time_listbox"
      aria-readonly={undefined}
      autoComplete="off"
      autoFocus={undefined}
      className="rw-input"
      disabled={undefined}
      id="rw_1_input"
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      readOnly={undefined}
      role="combobox"
      tabIndex={0}
      type="text"
      value=""
    />
    <span
      className="rw-select"
    >
      <button
        aria-disabled={false}
        aria-label="Select Date"
        className="rw-btn-calendar rw-btn"
        disabled={false}
        onClick={[Function]}
        tabIndex="-1"
        title="Select Date"
        type="button"
      >
        <span
          aria-hidden={true}
          className="rw-i rw-i-calendar"
        />
      </button>
      <button
        aria-disabled={false}
        aria-label="Select Time"
        className="rw-btn-time rw-btn"
        disabled={false}
        onClick={[Function]}
        tabIndex="-1"
        title="Select Time"
        type="button"
      >
        <span
          aria-hidden={true}
          className="rw-i rw-i-clock-o"
        />
      </button>
    </span>
  </div>
</div>
`;
